BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs

```diff
@@ -674,6 +676,44 @@ public void Flush()
             }
         }
 
+        /// <summary>
+        /// Asynchronously Flushes the in-memory buffer and any metrics being pre-aggregated.
+        /// </summary>
+        /// <returns>
+        /// True indicates telemetry data ownership is transferred out of process, that are emitted before the flush invocation.
+        /// False indicates transfer of telemetry data has failed, the process still owns all or part of the telemetry.
+        /// </returns>
+        /// <remarks>
+        /// <a href="https://go.microsoft.com/fwlink/?linkid=525722#flushing-data">Learn more</a>
+        /// </remarks>
+        public Task<bool> FlushAsync(CancellationToken cancellationToken)
+        {
+            if (this.TryGetMetricManager(out MetricManager privateMetricManager))
+            {
+                privateMetricManager.Flush(flushDownstreamPipeline: false);
+            }
+
+            TelemetryConfiguration pipeline = this.configuration;
+            if (pipeline != null)
+            {
+                MetricManager sharedMetricManager = pipeline.GetMetricManager(createIfNotExists: false);
+                sharedMetricManager?.Flush(flushDownstreamPipeline: false);
+
+                ITelemetryChannel channel = pipeline.TelemetryChannel;
+
+                if (channel is IAsyncFlushable asyncFlushableChannel && !cancellationToken.IsCancellationRequested)
+                {
+                    return asyncFlushableChannel.FlushAsync(cancellationToken);
+                }
+            }
+
+            return cancellationToken.IsCancellationRequested ? Task.Factory.StartNew(() =>
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                return false;
+            }, cancellationToken) : Task.FromResult(false);
```

Comment:
I presume this block of code is a substitute for `Task.FromCanceled(cancellationToken)`. Can you move it to a static utility class with `#ifdef`s so that targets that _do_ have Task.FromCanceled (all except NET45) can benefit from that built-in? I think we used to have a TaskEx class for things like that, but I can't see it in the sources.

Reply:
Added support for Task.Canceled and simplified return.